transport/controlbuf.go

```diff
@@ -365,6 +365,18 @@ func newLoopyWriter(s side, fr *framer, cbuf *controlBuffer, bdpEst *bdpEstimato
 const minBatchSize = 1000
 
 // run should be run in a separate goroutine.
+// It reads from the controlBuf in a loop and processes control frames
+// that came in by:
+// 1. Updating loopy's internal state, or/and
+// 2. Writing out HTTP2 frames on the wire.
+//
+// Loopy maintains all active streams with data to send in a linked-list.
+// All streams in the activeStreams linked-list must have both:
+// 1. Data to send, and
+// 2. Stream level flow control.
+//
+// In each iteration of rnu loop, other than processing the incoming control
```

Comment:
So, loopyWriter has a control buffer and active streams. How do they differ? I know the latter holds streams with data to send, but what does the control buffer hold?

(this is a question for me, not a passive-aggressive way to get you to update this comment - once I figure out the difference I may separately suggest an update to this)

Reply:
controlBuf is a data structure used to pass control frames to loopy by reader goroutine, stream goroutines and keepalive goroutine. 
activeStreams is loopy internal data structure to keep track of streams wanting to send data.

Reply:
- How do the two relate?
- Could you define "control frame"?
- I know that somewhere in here is round robin behavior as well as splitting messages that are too large. Where do these two happen? IIRC it was processData with the control frame..?

Reply:
loopy writer runs in a separate goroutine so other goroutines(read, streams, keepalive) need some way to tell it about:
1. Headers to send.
1. Data to send.
1. Incoming window updates
1. Incoming settings
1. Incoming goaways
1. Locally generated stream cleanup requests.

We use a data structure called `controlBuf`, which is made up of a lock and a list, to give this information to loopy. This information is passed as specific structs, we call them control frames, since they pass control information to loopy.
Following are the structs: https://github.com/grpc/grpc-go/blob/f1ab7acf3f0b8e779bc3592e7e9728cd2fd20402/transport/controlbuf.go#L87

Loopy can write window updates, acks, settings etc. in any order but data and headers of a stream need to be ordered, pass flow control and broken into data frames of max size 16KB(for now). To accomplish this loopy uses another data structure `activeStreams`(this is a list).

Reply:
Ok, thanks for this. I dug through the code some more. Some things we should clarify in documentation, I think:

Control buffer is the mechanism by which other goroutines communicate with loopy. Control buffer holds frames - data frames, header frames, and so on. Separately, loop holds a linked list of streams, and each stream holds a linked list of frames to send for that stream. Once loopy receives a frame, it enqueues it into a stream.

Later, as part of processData, loopy dequeues the next stream and peeks at the next frame to be sent. There is a limit to how much loopy can send each iteration; if the frame is too large, loopy will chop it up, send as much as it can, leave the rest of the frame to be sent next time loopy looks at the stream. If the frame is small enough, loopy sends it and dequeues it.

Since loopy iterates over a list of a streams, sending one piece of stream data at a time, loopy's scheduling is described as round robin across streams.

Reply:
I added documentation to controlBuffer, other than that the suggested documentation is already there (spread across run, processData and activeStream declaration).

The part about describing loopy's scheduling as round-robin doesn't seem necessary. 
1. We don't want to make a guarantee that it will stay this way. 
1. I don't see a user doing anything differently being equipped with this information.

What do you think?

Reply:
> I added documentation to controlBuffer

Checked it out - it looks great!

> We don't want to make a guarantee that it will stay this way.
> I don't see a user doing anything differently being equipped with this information.

I agree with these points, but I don't think the purpose of documenting this code is for the user - I think it's for the developer who works on this code. The extent to which you want your code to be approachable to other developers should be the guide. Given how large and important of an open source project gRPC is, I think it should be dialed pretty dialed in in terms of documentation and approachability.

> other than that the suggested documentation is already there

A developer could put together all the disparate pieces, and look at some code, and figure out the whole story. I think it'd be pretty useful to save that time and confusion by handing the whole story on a platter. How about this simplified version that does away with the flow parts?

> Loopy receives frames from the control buffer. Each frame is handled individually; most of the work done by loopy goes into handling data frames. Loopy maintains a queue of streams, and each stream maintains a queue of data frames; as loopy receives data frames it gets added to the queue of the relevant stream. The scheduling of data frames to be written involves round robin process during which loopy sends up to a maximum amount of data from one stream at a time, as stream and connection-level flow control permits.

Reply:
Looks great!!